### PR TITLE
SONARPY-552 Add ClassSymbol interface

### PR DESCRIPTION
**Pull request by guillaume-dequenne-sonarsource**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/semantic/ClassSymbolImpl.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.sonar.plugins.python.api.symbols.ClassSymbol;
+import org.sonar.plugins.python.api.symbols.Symbol;
+
+public class ClassSymbolImpl extends SymbolImpl implements ClassSymbol {
+
+  List<Symbol> parents = new ArrayList<>();
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/ClassSymbolImpl.java]: shouldn't they be private?

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java [hunk 2]
```diff
@@ -517,8 +521,55 @@ private void addSymbolUsage(Name nameTree) {
       SymbolImpl symbol = scope.resolve(nameTree.name());
       // TODO: use Set to improve performances
       if (symbol != null && symbol.usages().stream().noneMatch(usage -> usage.tree().equals(nameTree))) {
-        symbol.addUsage(nameTree, Usage.Kind.OTHER);
+          symbol.addUsage(nameTree, Usage.Kind.OTHER);
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java]: Unnecessary change?

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java [hunk 3]
```diff
@@ -517,8 +521,55 @@ private void addSymbolUsage(Name nameTree) {
...
       }
     }
+
+    private void resolveParents(ClassDef classDef) {
+      Symbol symbol = classDef.name().symbol();
+      if (symbol == null || !Symbol.Kind.CLASS.equals(symbol.kind())) {
+        return;
+      }
+      ClassSymbolImpl classSymbol = (ClassSymbolImpl) symbol;
+      if (hasMultipleBindingUsages(classSymbol)) {
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java]: I would suggest to change the kind of a classSymbol to OTHER inside `Scope#addBindingUsage`.
There you may add the following the code:
```java
if (symbol.kind() != Symbol.Kind.OTHER) {
  symbol.setKind(Symbol.Kind.OTHER);
}
```

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java [hunk 4]
```diff
@@ -517,8 +521,53 @@ private void addSymbolUsage(Name nameTree) {
...
+        if (!argument.is(Kind.REGULAR_ARGUMENT) || !(((RegularArgument) argument).expression() instanceof HasSymbol)) {
+          classSymbol.hasUnresolvedParents = true;
+          return;
+        }
+        Symbol parentSymbol = ((HasSymbol) ((RegularArgument) argument).expression()).symbol();
+        if (parentSymbol == null) {
+          classSymbol.hasUnresolvedParents = true;
+          return;
+        }
+        if (hasMultipleBindingUsages(parentSymbol)) {
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java]: I don't think we should have this check. We should add a symbol to parentSymbol even if it has multiple binding usages. It's then up to each rule to decide how to use those symbols.
Moreover setting `classSymbol.hasUnresolvedParents` is uncorrect here because we managed to find parent symbol (altough we don't know if it's a class or not.

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java [hunk 5]
```diff
@@ -517,8 +521,53 @@ private void addSymbolUsage(Name nameTree) {
...
+          classSymbol.hasUnresolvedParents = true;
+          return;
+        }
+        if (parentSymbol.fullyQualifiedName() != null && BuiltinSymbols.all().contains(parentSymbol.fullyQualifiedName())) {
+          classSymbol.addParent(parentSymbol);
+          continue;
+        }
+        if (!Symbol.Kind.CLASS.equals(parentSymbol.kind())) {
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java]: Even here, I think we should always add the found symbol and avoid setting `hasUnresolveParents` to true

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java (hunk 4 above)
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java]: I removed the check as it was made redundant by the `Scope#addBindingUsage` change, but as discussed, the `classSymbol.hasUnresolvedParents` will still be set to true in that case (and the parent won't be added to the list).

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolImpl.java [hunk 6]
```diff
@@ -39,6 +39,7 @@
   String fullyQualifiedName;
   private final List<Usage> usages = new ArrayList<>();
   private Map<String, Symbol> childrenSymbolByName = new HashMap<>();
+  private Kind kind;
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolImpl.java]: Shouldn't the kind be put to `OTHER`?

File: python-frontend/src/test/java/org/sonar/python/semantic/ClassSymbolTest.java [hunk 7]
```diff
@@ -20,112 +20,225 @@
...
+import static org.sonar.python.PythonTestUtils.parse;
 
 public class ClassSymbolTest {
+
+  @Test
+  public void no_parents() {
+    FileInput fileInput = parse(
+      "class C: ",
+      "  pass");
+    assertThat(fileInput.globalVariables()).hasSize(1);
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/test/java/org/sonar/python/semantic/ClassSymbolTest.java]: I don't understand the purpose of this test. Shouldn't you assert that `fileInput.globalVariables()` has 1 symbol of kind `CLASS` with empty parents property?

**Comment by andrea-guarino-sonarsource** [comment]
LGTM, please have a look at my 2 comments before merging
